amcl/test/basic_localization.py

```diff
@@ -21,61 +22,101 @@ def setUp(self):
-
-        return (a, abs(fmod(a - d_a + 5*pi, 2*pi) - pi))
+        return self.wrap_angle(a)
 
     def test_basic_localization(self):
         global_localization = int(sys.argv[1])
         self.target_x = float(sys.argv[2])
         self.target_y = float(sys.argv[3])
-        self.target_a = float(sys.argv[4])
+        self.target_a = self.wrap_angle_a(float(sys.argv[4]))
+        self.target_a = self.wrap_angle(self.target_a)
         tolerance_d = float(sys.argv[5])
         tolerance_a = float(sys.argv[6])
         target_time = float(sys.argv[7])
 
+        rospy.init_node('test', anonymous=True)
+        while rospy.rostime.get_time() == 0.0:
+            print('Waiting for initial time publication')
+            time.sleep(0.1)
+
         if global_localization == 1:
-            #print('Waiting for service global_localization')
+            print('Waiting for service global_localization')
             rospy.wait_for_service('global_localization')
             global_localization = rospy.ServiceProxy('global_localization', Empty)
             global_localization()
 
-        rospy.init_node('test', anonymous=True)
-        while(rospy.rostime.get_time() == 0.0):
-            #print('Waiting for initial time publication')
-            time.sleep(0.1)
         start_time = rospy.rostime.get_time()
-        # TODO: This should be replace by a pytf listener
-        rospy.Subscriber('/tf', TFMessage, self.tf_cb)
+        self.tfBuffer = tf2_ros.Buffer()
+        listener = tf2_ros.TransformListener(self.tfBuffer)
 
         while (rospy.rostime.get_time() - start_time) < target_time:
-            #print('Waiting for end time %.6f (current: %.6f)'%(target_time,(rospy.rostime.get_time() - start_time)))
+            print('Waiting for end time %.6f (current: %.6f)' % (target_time, (rospy.rostime.get_time() - start_time)))
+            self.get_pose()
             time.sleep(0.1)
 
-        (a_curr, a_diff) = self.compute_angle_diff()
-        print('Curr:\t %16.6f %16.6f %16.6f' % (self.tf.translation.x, self.tf.translation.y, a_curr))
-        print('Target:\t %16.6f %16.6f %16.6f' % (self.target_x, self.target_y, self.target_a))
-        print('Diff:\t %16.6f %16.6f %16.6f' % (
-            abs(self.tf.translation.x - self.target_x), abs(self.tf.translation.y - self.target_y), a_diff))
-        self.assertNotEquals(self.tf, None)
-        self.assertTrue(abs(self.tf.translation.x - self.target_x) <= tolerance_d)
-        self.assertTrue(abs(self.tf.translation.y - self.target_y) <= tolerance_d)
-        self.assertTrue(a_diff <= tolerance_a)
+        print("Final pose:")
+        self.get_pose()
+        a_curr = self.compute_angle()
+
+        self.assertNotEqual(self.tf, None)
+        self.assertAlmostEqual(self.tf.translation.x, self.target_x, delta=tolerance_d)
+        self.assertAlmostEqual(self.tf.translation.y, self.target_y, delta=tolerance_d)
+        self.assertAlmostEqual(a_curr, self.target_a, delta=tolerance_a)
```

Comment:
I think you can save a lot of headaches if you perform the wrapping on the difference between the angles. That way you don't need to force the angles to be within a certain range. It is a good result if the difference between the angles, wrapped to [-pi, pi], is almost 0.

Reply:
I think that should work as well. That is indeed maybe more elegant.